torch/nn/gradients.py

```diff
@@ -0,0 +1,209 @@
+"""Gradient interface"""
+
+import torch
+from .modules.utils import _pair, _triple
+
+
+def _output_padding(input, output_size, stride, padding, kernel_size):
+    output_size = list(output_size)
+    k = input.dim() - 2
+    print(k)
+    print(output_size)
+    if len(output_size) == k + 2:
+        output_size = output_size[-k:]
+    if len(output_size) != k:
+        raise ValueError("output_size must have {} or {} elements (got {})"
+                         .format(k, k + 2, len(output_size)))
+
+    def dim_size(d):
+        return ((input.size(d + 2) - 1) * stride[d] - 2 * padding[d] +
+                kernel_size[d])
+
+    min_sizes = [dim_size(d) for d in range(k)]
+    max_sizes = [min_sizes[d] + stride[d] - 1 for d in range(k)]
+    for size, min_size, max_size in zip(output_size, min_sizes, max_sizes):
+        if size < min_size or size > max_size:
+            raise ValueError(
+                ("requested an output size of {}, but valid sizes range "
+                 "from {} to {} (for an input of {})").format(
+                     output_size, min_sizes, max_sizes,
+                     input.size()[2:]))
+
+    return tuple([output_size[d] - min_sizes[d] for d in range(k)])
+
+
+def conv2d_input(in_channels, out_channels, out_backprop, weight, input_size,
+                 stride=1, padding=0, dilation=1, groups=1, bias=None):
+    r"""
+    Computes the gradient of conv2d with respect to the input of the convolution.
+    This is same as the 2D transposed convolution operator under the hood but requires
+    the shape of the gradient w.r.t. input to be specified explicitly.
+
+    Args:
+        in_channels (int): Number of channels in the input image
+        out_channels (int): Number of channels in the output gradient
+        out_backprop : output gradient tensor (minibatch x out_channels x oH x oW)
+        weight: filters tensor (out_channels x in_channels/groups x kH x kW)
+        input_size : Shape of the input gradient tensor
+        stride (int or tuple, optional): Stride of the convolution. Default: 1
+        padding (int or tuple, optional): Zero-padding added to both sides of the input. Default: 0
+        dilation (int or tuple, optional): Spacing between kernel elements. Default: 1
+        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
+        bias: optional bias tensor (out_channels). Default: None
+
+    Examples::
+
+        >>> out_backprop = autograd.Variable(torch.randn(1,1,3,2))
+        >>> weight = autograd.Variable(torch.randn(1,1,1,2))
+        >>> F.grad.conv2d_input(out_backprop, weight, stride=1, input_size=(1,1,3,3))
```

Comment:
A helpful thing to show is how `F.grad.conv2d_input` is related to a sequence of calls to `F.conv2d` and then `torch.autograd.grad()`. These would be good tests as well (which we need to have on the PR.)